segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -704,11 +718,162 @@ public void readTable(final WireCommands.ReadTable readTable) {
                                               .map(k -> new ByteArraySegment(k.getData().array()))
                                               .collect(Collectors.toList());
         tableStore.get(segment, keys, TIMEOUT)
-                  .thenAccept(values -> connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment, getTableEntries(keys, values))))
+                  .thenAccept(values -> connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment,
+                                                                                   getTableEntriesCommand(keys, values))))
                   .exceptionally(e -> handleException(readTable.getRequestId(), segment, operation, e));
     }
 
-    private WireCommands.TableEntries getTableEntries(final List<ArrayView> inputKeys, final List<TableEntry> resultEntries) {
+    @Override
+    public void getTableKeys(WireCommands.GetTableKeys getTableKeys) {
+        final String segment = getTableKeys.getSegment();
+        final String operation = "getTableKeys";
+
+        if (!verifyToken(segment, getTableKeys.getRequestId(), getTableKeys.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(getTableKeys.getRequestId(), "Fetching keys from {}.", getTableKeys);
+
+        int suggestedKeyCount = getTableKeys.getSuggestedKeyCount();
+        ByteBuffer token = getTableKeys.getContinuationToken();
+
+        byte[] state = null;
+        if (!token.equals(EMPTY_BYTE_BUFFER)) {
+            state = token.array();
+        }
+
+        final AtomicBoolean canContinue = new AtomicBoolean(true);
+        final AtomicInteger keyCount = new AtomicInteger(0);
+        final AtomicInteger msgSize = new AtomicInteger(0);
+        final AtomicReference<ByteBuffer> continuationToken = new AtomicReference<>(EMPTY_BYTE_BUFFER);
+        final LinkedBlockingQueue<TableKey> keys = new LinkedBlockingQueue<>();
+        tableStore.keyIterator(segment, state, TIMEOUT)
+                  .thenCompose(itr -> Futures.loop(
+                          canContinue::get,
+                          itr::getNext,
+                          e -> {
+                              if (e == null) {
+                                  canContinue.set(false);
+                              } else {
+                                  // store all keys.
+                                  keys.addAll(e.getEntries());
+                                  // update the continuation token.
+                                  ArrayView lastState = e.getState();
+                                  continuationToken.set(ByteBuffer.wrap(lastState.array(), lastState.arrayOffset(), lastState.getLength()));
+                                  // update the keycount and msgSize.
+                                  keyCount.addAndGet(e.getEntries().size());
+                                  msgSize.addAndGet(getTableKeyBytes(segment, e));
```

Comment:
This and line 759 cause the items to be iterated over twice. We should combine these into a single run through the iterator.

Reply:
refactored it.